src/main/java/org/opensearch/knn/training/VectorReader.java

```diff
@@ -182,10 +184,20 @@ public void onResponse(SearchResponse searchResponse) {
             int vectorsToAdd = Integer.min(maxVectorCount - collectedVectorCount, hits.length);
             List<Float[]> trainingData = new ArrayList<>();
 
-            for (int i = 0; i < vectorsToAdd; i++) {
-                trainingData.add(
-                    ((List<Number>) hits[i].getSourceAsMap().get(fieldName)).stream().map(Number::floatValue).toArray(Float[]::new)
-                );
+            for (int vector = 0; vector < vectorsToAdd; vector++) {
+                Map<String, Object> originalSourceMap = hits[vector].getSourceAsMap();
+                // The field name may be a nested field, so we need to split it and traverse the map.
+                // Example fieldName: "my_field" or "my_field.nested_field.nested_nested_field"
+                String[] fieldPath = fieldName.split("\\.");
+                Map<String, Object> currentMap = deepCopyMap(originalSourceMap);
```

Comment:
Why do we need a deep copy just to extract the vector from the map?

Reply:
In case we going to mutate the map content if will affect the original map, it's the previous feedback from @martin-gaievski https://github.com/opensearch-project/k-NN/pull/1318#discussion_r1409733595

Reply:
Why do we need to mutate the map?

Reply:
Separated out this part to a function and used the original source map directly